### PR TITLE
Karinna Iniguez - Ride Share - Octos

### PR DESCRIPTION
**Pull request by karinnainiguez**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I had to decide where my new methods (including helper methods) belonged.  I made each decision by asking myself, "If I asked these questions of the instance, would it make sense?  Or would they have to go ask another class instance?".  Organizing both my classes and the tests for my classes brought me back to that question.  All the class instances interact with each other, so thinking about which class or test file should hold that code was messy and difficult for me.
Describe any examples of composition that you encountered in this project, if any |  The trip dispatcher has a collection of trips, a collection of passenger instances, and a collection of drivers.  Each driver instance and passenger instance also hold a collection of trips.  Finally, trips have an instance of passenger, and an instance of driver.
Describe a _nominal test_ that you wrote for this assignment. |  To test my Driver#total_revenue method, I wrote a test that took an instance of driver that contained three completed trips with costs > 0.  The test expected a total revenue that equalled the three costs (minus 1.65 fee from each) times the percentage that the drivers get (80%). 
Describe an _edge case test_ that you wrote for this assignment. | For that same Driver#total_revenue I also wrote tests to make sure that the method returned 0 if the driver did not have any trips, or (another edge case test) if the trips were less than the fee charged.
Describe a concept that you gained more clarity on as you worked on this assignment. | I was able to better understand how instances of different classes can interact with has-a relationships.  Although they did not inherit anything from other classes, they each contained and interacted with each other.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -19,29 +22,77 @@ def initialize(input)
       @status = input[:status] == nil ? :AVAILABLE : input[:status]
 
       @trips = input[:trips] == nil ? [] : input[:trips]
-    end
+    end # initialize
+
+    def completed_trips
+      @trips.find_all {|trip| trip.end_time != nil}
+    end # completed_trips
```
Comment (droberts-sea) [inline lib/driver.rb:29]: I love that you've broken this out as a separate helper method! This makes the other code in this file much easier to read.

File: lib/trip_dispatcher.rb [hunk 2]
```diff
@@ -87,15 +95,77 @@ def load_trips
...
+    def next_driver
+      # Array of available drivers
+      all_available = @drivers.find_all do |driver|
+        # Minnie (ID 100) is avail, and no current trips,
+        # but assignment doesnt want us to choose her.
+        driver.trips.length > 0 && driver.status == :AVAILABLE
+      end
+
+      return nil if all_available == []
+
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:110]: I like the choice to return `nil` here, instead of raising an exception. This gives whoever called this method the power to raise an exception if needed, but doesn't force it.

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -87,15 +95,77 @@ def load_trips
...
+      last_trips = all_available.map do |driver|
+        driver.trips.min_by do |trip|
+          Time.now - trip.end_time
+        end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:114]: Good use of enumerables here

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -76,5 +77,93 @@
...
+    it "does not return a negative revenue" do
+      trips = [RideShare::Trip.new({cost:1.30, rating: 3}), RideShare::Trip.new({cost:1.25, rating: 3}), RideShare::Trip.new({cost:1.03, rating: 3})]
+      driver = RideShare::Driver.new({id:56, vin:"a"*17, trips: trips})
+      expected_revenue = 0
+      result = driver.total_revenue
+      result.must_equal expected_revenue
+    end
+
+    it "accurately ignores trips in progress" do
+      start_time = Time.parse("2016-02-16T12:45:00+00:00")
```
Comment (droberts-sea) [inline specs/driver_spec.rb:126]: Good job fleshing out the interesting test cases for all these wave 1 methods.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | yes
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | yes
Complete tests | yes
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | yes
Appropriate use of Ruby's `Date` | yes
Appropriate helper methods were made to help with complex logic | yes
Complete tests | yes
**Overall** | Excellent job overall! Your code is clear and easy to understand, your logic is sound, and your test coverage is superb. I have left a few nitpicks below, but in general I am quite pleased with this submission, and excited to see what you come up with on Hotel. Keep up the hard work!
